File: plugins/agent-skills/skills/claude-delegate/scripts/claude_delegate/artifact_lifecycle.py
from __future__ import annotations

import fcntl
import gzip
import shutil
from datetime import datetime, timedelta, timezone
from pathlib import Path

from .common import TERMINAL_JOB_STATES, artifact_paths, read_json, utc_now, write_json
from .ledger import collect_summaries, parse_timestamp


COMPACTABLE_JOB_FIELDS = ("events_path", "stdout_path", "stderr_path")


def _gzip_path(path: Path) -> Path:
    gz_path = path.with_name(f"{path.name}.gz")
    with path.open("rb") as src, gzip.open(gz_path, "wb") as dst:
        shutil.copyfileobj(src, dst)
    path.unlink()
    return gz_path


def _preserved_artifacts(paths: dict[str, Path]) -> dict[str, str]:
    return {
        "request_path": str(paths["request"]),
        "job_metadata_path": str(paths["job"]),
        "normalized_path": str(paths["normalized"]),
        "handoff_path": str(paths["handoff"]),
        "patch_path": str(paths["patch"]),
    }


def compact_job_artifacts(job_path: str | Path) -> dict:
    artifacts_dir = Path(job_path)
    paths = artifact_paths(artifacts_dir)
    job_lock = paths["lock"]
    if not job_lock.exists():
        return {
            "ok": False,
            "error_type": "missing_job_state",
            "error_message": "job lock file does not exist",
            "job_path": str(artifacts_dir),
        }

    with job_lock.open("r+") as lock_file:
        fcntl.flock(lock_file, fcntl.LOCK_EX)

        job = read_json(paths["job"]) or {}
        if job.get("state") not in TERMINAL_JOB_STATES:
            return {
                "ok": False,
                "error_type": "non_terminal_job",
                "error_message": "job is not terminal",
                "job_path": str(artifacts_dir),
            }

        updated_paths: dict[str, str] = {}
        compacted_fields: list[str] = []

        for field in COMPACTABLE_JOB_FIELDS:
            raw_path = job.get(field)
            if not raw_path:
                continue
            current_path = Path(raw_path)
            if current_path.suffix == ".gz" or not current_path.exists():
                continue
            gz_path = _gzip_path(current_path)
            updated_paths[field] = str(gz_path)
            compacted_fields.append(field)

        if not compacted_fields:
            metadata = job.get("artifact_lifecycle") or {}
            if metadata:
                return {
                    "ok": True,
                    "job_path": str(artifacts_dir),
                    "status": metadata.get("status", "unchanged"),
                    "compacted_fields": metadata.get("compacted_fields", []),
                }
            return {
                "ok": True,
                "job_path": str(artifacts_dir),
                "status": "unchanged",
                "compacted_fields": [],
            }

        metadata = {
            "status": "compacted",
            "compacted_at": utc_now(),
            "compacted_fields": compacted_fields,
            "compacted_paths": updated_paths,
            "preserved_artifacts": _preserved_artifacts(paths),
        }

        job.update(updated_paths)
        job["artifact_lifecycle"] = metadata
        job["updated_at"] = metadata["compacted_at"]
        write_json(paths["job"], job)

        return {
            "ok": True,
            "job_path": str(artifacts_dir),
            "status": "compacted",
            "compacted_fields": compacted_fields,
            "compacted_paths": updated_paths,
        }


def compact_terminal_jobs(
    artifacts_root: str | Path,
    *,
    older_than_hours: float,
) -> dict:
    cutoff = datetime.now(timezone.utc) - timedelta(hours=older_than_hours)
    items, _ = collect_summaries(
        artifacts_root,
        limit=None,
        session_id=None,
        state=None,
    )

    results: list[dict] = []
    for item in items:
        finished_at = parse_timestamp(item.get("finished_at"))
        if item.get("state") not in TERMINAL_JOB_STATES:
            continue
        if finished_at is None or finished_at > cutoff:
            continue
        results.append(compact_job_artifacts(item["job_path"]))

    compacted = [item for item in results if item.get("status") == "compacted"]
    return {
        "ok": True,
        "count": len(results),
        "compacted_count": len(compacted),
        "items": results,
        "cutoff": cutoff.isoformat(),
    }

File: plugins/agent-skills/skills/claude-delegate/src/lib/artifact-lifecycle.ts
import {
  closeSync,
  createReadStream,
  createWriteStream,
  existsSync,
  openSync,
  unlinkSync,
} from "node:fs";
import { createGzip } from "node:zlib";
import { pipeline } from "node:stream/promises";

import { flockSync } from "fs-ext";

import {
  TERMINAL_JOB_STATES,
  artifactPaths,
  readJson,
  utcNow,
  writeJson,
} from "./common.js";
import { collectSummaries, parseTimestamp } from "./ledger.js";

const COMPACTABLE_JOB_FIELDS = [
  "events_path",
  "stdout_path",
  "stderr_path",
] as const;

async function gzipPath(filePath: string): Promise<string> {
  const gzPath = `${filePath}.gz`;
  await pipeline(
    createReadStream(filePath),
    createGzip(),
    createWriteStream(gzPath),
  );
  unlinkSync(filePath);
  return gzPath;
}

function preservedArtifacts(
  paths: ReturnType<typeof artifactPaths>,
): Record<string, string> {
  return {
    request_path: paths.request,
    job_metadata_path: paths.job,
    normalized_path: paths.normalized,
    handoff_path: paths.handoff,
    patch_path: paths.patch,
  };
}

export async function compactJobArtifacts(
  jobPath: string,
): Promise<Record<string, unknown>> {
  const paths = artifactPaths(jobPath);
  if (!existsSync(paths.lock)) {
    return {
      ok: false,
      error_type: "missing_job_state",
      error_message: "job lock file does not exist",
      job_path: jobPath,
    };
  }

  const fd = openSync(paths.lock, "r+");
  try {
    flockSync(fd, "ex");

    const job = readJson(paths.job) ?? {};
    if (!TERMINAL_JOB_STATES.has(job.state as string)) {
      return {
        ok: false,
        error_type: "non_terminal_job",
        error_message: "job is not terminal",
        job_path: jobPath,
      };
    }

    const updatedPaths: Record<string, string> = {};
    const compactedFields: string[] = [];

    for (const field of COMPACTABLE_JOB_FIELDS) {
      const rawPath = job[field] as string | undefined;
      if (!rawPath) continue;
      if (rawPath.endsWith(".gz") || !existsSync(rawPath)) continue;
      const gzPath = await gzipPath(rawPath);
      updatedPaths[field] = gzPath;
      compactedFields.push(field);
    }

    if (compactedFields.length === 0) {
      const metadata = (job.artifact_lifecycle ?? {}) as Record<
        string,
        unknown
      >;
      if (Object.keys(metadata).length > 0) {
        return {
          ok: true,
          job_path: jobPath,
          status: (metadata.status as string) ?? "unchanged",
          compacted_fields:
            (metadata.compacted_fields as string[]) ?? [],
        };
      }
      return {
        ok: true,
        job_path: jobPath,
        status: "unchanged",
        compacted_fields: [],
      };
    }

    const metadata: Record<string, unknown> = {
      status: "compacted",
      compacted_at: utcNow(),
      compacted_fields: compactedFields,
      compacted_paths: updatedPaths,
      preserved_artifacts: preservedArtifacts(paths),
    };

    Object.assign(job, updatedPaths);
    job.artifact_lifecycle = metadata;
    job.updated_at = metadata.compacted_at;
    writeJson(paths.job, job);

    return {
      ok: true,
      job_path: jobPath,
      status: "compacted",
      compacted_fields: compactedFields,
      compacted_paths: updatedPaths,
    };
  } finally {
    flockSync(fd, "un");
    closeSync(fd);
  }
}

export async function compactTerminalJobs(
  artifactsRoot: string,
  opts: { olderThanHours: number },
): Promise<Record<string, unknown>> {
  const cutoff = new Date(
    Date.now() - opts.olderThanHours * 3600_000,
  );
  const [items] = collectSummaries(artifactsRoot, {
    limit: null,
    sessionId: null,
    runtime: null,
    provider: null,
    state: null,
  });

  const results: Record<string, unknown>[] = [];
  for (const item of items) {
    const finishedAt = parseTimestamp(
      item.finished_at as string | null,
    );
    if (!TERMINAL_JOB_STATES.has(item.state as string)) continue;
    if (
      finishedAt === null ||
      finishedAt.getTime() > cutoff.getTime()
    )
      continue;
    results.push(await compactJobArtifacts(item.job_path as string));
  }

  const compacted = results.filter(
    (item) => item.status === "compacted",
  );
  return {
    ok: true,
    count: results.length,
    compacted_count: compacted.length,
    items: results,
    cutoff: cutoff.toISOString(),
  };
}
